pkg/canary/comparator/comparator.go

```diff
@@ -0,0 +1,273 @@
+package comparator
+
+import (
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/grafana/loki/pkg/canary/reader"
+)
+
+const (
+	ErrOutOfOrderEntry    = "out of order entry %s was received before entries: %v\n"
+	ErrEntryNotReceivedWs = "websocket failed to receive entry %v within %f seconds\n"
+	ErrEntryNotReceived   = "failed to receive entry %v within %f seconds\n"
+	ErrDuplicateEntry     = "received a duplicate entry for ts %v\n"
+	ErrUnexpectedEntry    = "received an unexpected entry with ts %v\n"
+)
+
+var (
+	totalEntries = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "loki_canary",
+		Name:      "total_entries",
+		Help:      "counts log entries written to the file",
+	})
+	outOfOrderEntries = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "loki_canary",
+		Name:      "out_of_order_entries",
+		Help:      "counts log entries received with a timestamp more recent than the others in the queue",
+	})
+	wsMissingEntries = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "loki_canary",
+		Name:      "websocket_missing_entries",
+		Help:      "counts log entries not received within the maxWait duration via the websocket connection",
+	})
+	missingEntries = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "loki_canary",
+		Name:      "missing_entries",
+		Help:      "counts log entries not received within the maxWait duration via both websocket and direct query",
+	})
+	unexpectedEntries = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "loki_canary",
+		Name:      "unexpected_entries",
+		Help:      "counts a log entry received which was not expected (e.g. received after reported missing)",
+	})
+	duplicateEntries = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: "loki_canary",
+		Name:      "duplicate_entries",
+		Help:      "counts a log entry received more than one time",
+	})
+	responseLatency prometheus.Histogram
+)
+
+type Comparator struct {
+	entMtx        sync.Mutex
+	w             io.Writer
+	entries       []*time.Time
+	ackdEntries   []*time.Time
+	maxWait       time.Duration
+	pruneInterval time.Duration
+	sent          chan time.Time
+	recv          chan time.Time
+	rdr           reader.LokiReader
+	quit          chan struct{}
+	done          chan struct{}
+}
+
+func NewComparator(writer io.Writer, maxWait time.Duration, pruneInterval time.Duration,
```

Comment:
This is definitely a nitpick but `comparator.NewComparator`, `reader.NewReader` and `writer.NewWriter` would annoy some go linters and would have `comparator.New`, `reader.New` and `writer.New` be suggested instead. I'm ok with leaving it as it is, but we do have other packages that just use `New` (loki, distributor, ingester, stage, promtail, etc)

Reply:
Good feedback, i'll follow up with a separate PR for this!